### PR TITLE
Switch calls to use headers and add apikey auth

### PR DESCRIPTION
**Pull request by ROMzombie**
We shouldn't be sending tokens via the URI (which can be logged or captured in transit), so switch to using HTTP Headers instead for all calls.

Add support for the `X-Keap-API-Key` header to allow Personal Access Tokens and Service Account Keys through the SDK.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,6 +39,11 @@ Or manually add it to your composer.json:
 
 ## Authentication
 
+Currently Keap supports two types of authentication for our APIs: the OAuth2 Access Code Grant and API Keys.  
+Developers of third-party integrations should always use our OAuth2 authentication, but developers building integrations for a single tenant may find the use of API Keys much simpler.
+
+### OAuth2 Access Code Grant
+
 The client ID and secret are the key and secret for your OAuth2 application found at the [Infusionsoft Developers](https://keys.developer.infusionsoft.com/apps/mykeys) website.
```
Comment (mickaelsavafi-keap) [inline README.md:47]: (not part of your changes but..) Broken link?

File: README.md [hunk 2]
```diff
@@ -75,6 +80,26 @@ if ($infusionsoft->getToken()) {
 }
 ```
 
+### API Keys
+
+API Keys are a "password" for your data in an application and should always be treated like a dangerous secret. 
+
+In our UI you will find an API Settings screen which divides API Keys into two distinct categories:
+* Personal Access Tokens, which are scoped to your own user account and can only see and manipulate the data you have access to
```
Comment (mickaelsavafi-keap) [inline README.md:88]:
```suggestion
* `Personal Access Tokens`, which are scoped to your own user account and can only see and manipulate the data you have access to.
```

File: README.md [hunk 3]
```diff
@@ -75,6 +80,26 @@ if ($infusionsoft->getToken()) {
...
+* Service Account Keys, which can only be authorized by an Admin and have full access to the data stored in the application
```
Comment (mickaelsavafi-keap) [inline README.md:89]:
```suggestion
* `Service Account Keys`, which can only be authorized by an Administrator and have full access to the data stored in the application.
```

File: README.md [hunk 4]
```diff
@@ -75,6 +80,26 @@ if ($infusionsoft->getToken()) {
...
+
+For additional information on how to authorize and use PATs and SAKs please see our [developer documentation](https://developer.infusionsoft.com/pat-and-sak/).
```
Comment (mickaelsavafi-keap) [inline README.md:91]:
```suggestion
For additional information on how to authorize and use PATs and SAKs please see our [developer documentation](https://developer.keap.com/pat-and-sak/).
```

Comment (mickaelsavafi-keap) [reply README.md:91]: Those seem to be slightly different? WDYT? Can we use that?

File: src/Infusionsoft/Infusionsoft.php [hunk 5]
```diff
@@ -433,27 +475,40 @@ public function request()
...
-            $params = array_merge(array('key' => $token->getAccessToken()), $params);
+            $params = array_merge(array('key' => 'backwards-compatability'), $params);
         }
 
         // Reset the empty key flag back to the default for the next request
         $this->needsEmptyKey = true;
 
+        $options = [];
+
+        if ( $this->authenticationType === AuthenticationType::OAuth2AccessToken ) {
```
Comment (mickaelsavafi-keap) [inline src/Infusionsoft/Infusionsoft.php:497]: Since the same code is being used below (line 542-550), should this be a function?

**Review by mickaelsavafi-keap — APPROVED** [review]

File: README.md (hunk 4 above)
Comment (ROMzombie) [reply README.md:91]: It's part of the pain and weirdness of how the developer sites are set up right now on Wordpress.  You can use either for direct links due to the nginx rewrite rules, but internal redirects will end up on the Infusionsoft version until we get a rework done.

File: src/Infusionsoft/Http/InfusionsoftHttpAdapterTransport.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /**
+     * @param $endpoint
+     * @param $payload
+     * @return string
+     * @throws \Psr\Http\Client\ClientExceptionInterface
+     */
+    public function send($endpoint, $payload)
+    {
```
[the remark this answers is not in the document]
Comment (ROMzombie) [reply src/Infusionsoft/Http/InfusionsoftHttpAdapterTransport.php:36]: This is a reimplementation of the fXmlRpc HttpAdapterTransport with the extra parameter for default options added to allow configuration of the headers on the request.   The original didn't enforce it, so I didn't bother getting any more explicit. 🤷‍♂️

File: src/Infusionsoft/Infusionsoft.php [hunk 7]
```diff
@@ -433,27 +475,40 @@ public function request()
...
-
         $params = func_get_args();
         $method = array_shift($params);
 
         // Some older methods in the API require a key parameter to be sent
         // even if OAuth is being used. This flag can be made false as it
         // will break some newer endpoints.
         if ($this->needsEmptyKey) {
-            $params = array_merge(array('key' => $token->getAccessToken()), $params);
+            $params = array_merge(array('key' => 'backwards-compatability'), $params);
```
Comment (ROMzombie) [inline src/Infusionsoft/Infusionsoft.php:489]: This is one of the weird points; as far as I'm aware, the API authentication filter _checks_ for a value there, but the actual value doesn't matter, so I changed it to a dummy string instead.

File: README.md (hunk 4 above)
Comment (mickaelsavafi-keap) [reply README.md:91]: 😢 Makes sense, sadly!!!

File: src/Infusionsoft/Http/InfusionsoftHttpAdapterTransport.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (mickaelsavafi-keap) [reply src/Infusionsoft/Http/InfusionsoftHttpAdapterTransport.php:36]: Makes sense, all good then.

**Pull request merged**